### PR TITLE
Support case-insensitive substring def matching

### PR DESCRIPTION
**Pull request by stuarth**
defs_for_name / name_defs replacements that match more broadly.

Supports https://github.com/rust-lang-nursery/rls/issues/514

### CODE REVIEW

**Comment by nrc** [comment]
There seems to be test failure on the CI. Do you see that locally?

File: src/lib.rs [hunk 1]
```diff
@@ -350,17 +350,17 @@ impl<L: AnalysisLoader> AnalysisHost<L> {
         })
     }
 
-    pub fn name_defs(&self, name: &str) -> AResult<Vec<Def>> {
+    pub fn matching_defs(&self, query: &str) -> AResult<Vec<Def>> {
```
Comment (nrc) [inline src/lib.rs]: I would prefer `substr` to `query` as an arg. Could you add a comment explaining how this matches (substring, case-insensitive) please?

File: src/analysis.rs [hunk 2]
```diff
@@ -209,13 +209,19 @@ impl Analysis {
...
+    pub fn defs_matching_query(&self, query: &str) -> Vec<Def> {
+        let lowered_query = query.to_lowercase();
+
         self.for_all_crates(|c| {
-            c.def_names.get(name).map(|ids| {
-                ids.into_iter()
-                    .flat_map(|id| c.defs.get(id).cloned())
-                    .collect()
-            })
+            let defs: Vec<Def> = c.def_names.iter().fold(vec![], |mut acc, (ref def_n, ref ids)| {
```
Comment (nrc) [inline src/analysis.rs]: I would use filter and flat_map, rather than fold, I think the code would be simpler and more efficient.

**Comment by stuarth** [comment]
Thanks a bunch for the feedback @nrc ! Appreciate you taking the time!

re performance, this version is quite a bit slower. Super simple benchmarking (i.e. call the fn 1000x and average the results) in the rls crate shows about a 100x increase in speed. Roughly half of that increase comes from scanning each key (instead of a hashmap lookup), and the rest lower casing and the `.contains` call. 

That said, it's a quick call in absolute terms (a couple ms on my machine) and the broader result set is useful. For me it's a good trade off, especially as it's a call that's not done too frequently.

What do you think?

**Comment by nrc** [comment]
>  I do believe it'd be wise to build a separate fuzzy index/lookup tree on the side, since it can be considerably quicker to query that instead ...

I agree. However, I'm not actually sure how to do that with the same semantics as in this PR. It would be easier to match just a prefix (I think using a trie for this would be the best option), rather than an arbitrary substring (and I think just as useful). I'm also not sure how to efficiently implement case-insensitive prefix matching (at least in a way that is fully unicode-compliant).

So, I think it would be OK to land this PR as is, and keep an open issue for a more efficient implementation. Unless @stuarth wants to implement a more efficient version now?

**Comment by stuarth** [comment]
Let's switch to prefix matching. Completely agree about the performance concerns, and it's probably a bad idea to offer semantics that we can't fulfill efficiently.

@nrc, re 
> I'm also not sure how to efficiently implement case-insensitive prefix matching (at least in a way that is fully unicode-compliant).

curious if you could spell that out for me a bit 😄 . If we added lower-cased symbol names to a trie for prefix matching, that wouldn't be unicode compliant?

**Comment by stuarth** [comment]
Also, 1. Thank you both for the feedback! 2. Assuming we want to back the implementation with a trie, it looks like we'll be looking to an external crate. Any preferences?

**Comment by nrc** [comment]
I don't know of any crate to do this, sorry.

I was assuming that running to_lowercase would be too slow, but given you'd only do it on the search term, (and you'd need to run it in any case on the symbols) I think that is not too bad an approach.

**Comment by stuarth** [comment]
I reimplemented with a radix trie (and limited scope to starts with matching) and the matching performance is good -- nearly identical to the original version's exact string lookup

**Comment by stuarth** [comment]
This approach does add a cost when we're indexing the Defs. On my machine that's ~0.5 seconds on the initial indexing for all crates depended on by rls, and ~0.05 seconds for incrementally indexing rls

**Comment by nrc** [comment]
> On my machine that's ~0.5 seconds on the initial indexing for all crates depended on by rls

That seems tolerable. In general, I'm keen to keep this number low because it has an impact on the user experience, but this seems a worthwhile cost.

**Comment by stuarth** [comment]
Sounds good. The radix-trie crate does have a [fairly large set of tests](https://github.com/michaelsproul/rust_radix_trie/blob/master/src/test.rs). I added a [test](https://github.com/nrc/rls-analysis/pull/118/commits/ea060e8b2df3788d3969bf64d2d35f88be7fb38f) matching various node types. Any additional tests you'd like to see?

BTW, looks like the travis failure is a travis hiccup rather than test failure. Any chance I could get you to rerun it @nrc ?

**Comment by nrc** [comment]
I think a few more tests along the lines of the one you added would be fine. It would be good to have a longer search string as well as the short one, a search that should only return an exact match, one which should return zero matches, and a search which includes a unicode character.

**Comment by nrc** [comment]
Tests restarted

**Comment by stuarth** [comment]
Thanks once again the feedback and patience @nrc ! I think we're good now, but please let me know if you'd like to see anything else.

Also, I added unicode tests to the trie lib https://github.com/michaelsproul/rust_radix_trie/commit/9cda4a0325cad5b8ca4137d3db7bbd214f4f7df0

**Comment by nrc** [comment]
Great! Thanks for keeping going with this!

**Comment by nrc** [comment]
part of 0.9.3

**Pull request merged**